lib/Corres_UL.thy

```diff
@@ -1178,6 +1178,18 @@ lemma corres_move_asm:
 
 lemmas corres_cross_over_guard = corres_move_asm[rotated]
 
+lemma corres_cross_add_guard:
```

Comment:
Note: these two lemmas can be expressed using cross_rel. But I don't know if that's necessary. Just wanted to make the observation in case it's useful.

Reply:
It is actually slightly different, I think. `cross_rel` only uses `P` and the state_relation to derive `Q`, but here `Q` also depends on `P’`. This dependency exists in and is copied from the `corres_cross_over_guard` lemma defined just above here. I am not sure, but maybe `cross_rel` is slightly restrictive than what we always want. Or, this `P’` is not often used and hence we could do without most of the time. It might be worth while to investigate, and depending how it goes we could consider converting existing some of the cross lemmas into `cross_rel` statements.

Reply:
I should have been clearer, and I should check in an isabelle session, but I think the cross_rel expression would be `cross_rel P (%s'. P' s' -> Q' s') `

Reply:
But this trick won't work on the second lemma. So it's true I haven't completely thought this through.

Reply:
Ah, you are right. That second one is in the opposite direction. And,

> I should have been clearer, and I should check in an isabelle session, but I think the cross_rel expression would be `cross_rel P (%s'. P' s' -> Q' s')`

yes, that does work! I’d find the expression slightly less intuitive in that Q’ actually depends on both P and P’ (probably more on P), but it’s rare that we need to use P’ anyway. This PR is merged now, but it would be really good to start using `cross_rel` more, I agree. There is [JIRA-1290](https://sel4.atlassian.net/browse/VER-1290) for this which we should get on to when we can.

Reply:
And, just to mention, this `corres_cross_add_guard` is something I added a while ago, and is used in several places including Michael’s `add_sym_refs` method. What is new is the one in the other direction (the one next to this). I moved both to lib on this occasion.